### ISSUE TITLE
[sival/earlgrey-entropy_src] ENTROPY_SRC Test do not progress or fails due to Threshold Settings

### ISSUE THREAD

**Issue by github-gcontributor** [lowRISC/opentitan#25720]
### Test point name

[chip_sw_entropy_src_fips_mode_health_tests]https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/data/ip/chip_entropy_src_testplan.hjson#L144

### Host side component

None

### OpenTitanTool infrastructure implemented

None

### Silicon Validation (SiVal)

Yes

### Emulation Targets

- [ ] None
- [ ] CW310
- [x] Hyperdebug + CW310

### Contact person

vogelpi 

### Checklist

I am running the test "entropy_src_fips_mode_health_test" (PR: https://github.com/lowRISC/opentitan/pull/25719) that checks the ENTROPY_SRC module’s behavior under different health test thresholds. The test involves enabling ENTROPY_SRC in FIPS mode, configuring CSRNG and EDNs, and eventually triggering a recoverable alert (Step 16).

However, I am facing the following problem:

When I set the health test thresholds to very low (extremely stringent) values, the ENTROPY_SRC never passes its health checks. Because of this, no valid entropy is produced, and the EDNs and CSRNG cannot start operating as expected. The test gets stuck in EDN initialization phase.
If I loosen the thresholds (for example, using values like 0x1FF or 0x1F0), the ENTROPY_SRC does produce valid entropy, and EDNs and CSRNG initialize successfully. But with these more lenient thresholds, the ENTROPY_SRC does not fail later as intended, so the recoverable alert bit (expected at Step 16) never gets set.
To summarize:
While varying the stringent test parameters to extreme values, causes the EDN to get stuck and loosening the values (that passes the EDN configuration) results in failure of the function to verify the RECOVERABLE ALERT bit set.

Can this be a configuration issue? Can you please suggest your inputs?
Any advice on threshold values or test configuration changes or missing steps that would allow to generate recoverable alert test function to pass would be greatly appreciated.

Trials:
.high_threshold = 0x1FF, --> proceeding for all health parameters
.high_threshold = 0x1,  --> EDN stucks in infinite loop for all healthparameters 
.high_threshold = 0x2,  --> EDN stucks in infinite loop for all health parameters 
.high_threshold = 0xf,  --> EDN stucks in infinite loop for all health parameters 
.high_threshold = 0xff,  --> EDN stucks in infinite loop for all health parameters 
.high_threshold = 0x10F,  --> EDN stucks in infinite loop for all health parameters 
.high_threshold = 0x1F0,--> proceeding for all health parameters 
.high_threshold = 0x10A  --> EDN stucks in infinite loop, for all health parameters 
.high_threshold = 0x1FA,--> proceeding for all health parameters 
.high_threshold = 0x10E  --> EDN stucks in infinite loop, for all health parameters

Platform:
CW340+Hyperdebug

BUILD file changes (not yet checked in):
opentitan_test(
    name = "entropy_src_fips_mode_health_test",
    srcs = ["entropy_src_fips_mode_health_test.c"],
    exec_env = dicts.add(
        EARLGREY_TEST_ENVS,
        EARLGREY_SILICON_OWNER_ROM_EXT_ENVS,
        {
            "//hw/top_earlgrey:fpga_cw310_sival": None,
            "//hw/top_earlgrey:sim_verilator": None,
            "//hw/top_earlgrey:fpga_cw340_sival": None,
            "//hw/top_earlgrey:silicon_creator": None,
        },
    ),
    verilator = verilator_params(timeout = "long"),
    deps = [
        ":otbn_randomness_impl",
        "//hw/top_earlgrey/sw/autogen:top_earlgrey",
        "//sw/device/lib/base:memory",
        "//sw/device/lib/base:mmio",
        "//sw/device/lib/dif:aes",
        "//sw/device/lib/dif:alert_handler",
        "//sw/device/lib/dif:entropy_src",
        "//sw/device/lib/dif:keymgr",
        "//sw/device/lib/dif:kmac",
        "//sw/device/lib/dif:otbn",
        "//sw/device/lib/dif:otp_ctrl",
        "//sw/device/lib/dif:pwrmgr",
        "//sw/device/lib/dif:rv_core_ibex",
        "//sw/device/lib/runtime:hart",
        "//sw/device/lib/runtime:log",
        "//sw/device/lib/testing:aes_testutils",
        "//sw/device/lib/testing:alert_handler_testutils",
        "//sw/device/lib/testing:entropy_testutils",
        "//sw/device/lib/testing:keymgr_testutils",
        "//sw/device/lib/testing:otp_ctrl_testutils",
        "//sw/device/lib/testing:pwrmgr_testutils",
        "//sw/device/lib/testing/test_framework:ottf_main",
    ],
)
Please fill out this checklist as items are completed. Link to PRs and issues as appropriate.
- [ ] Check if existing test covers most or all of this testpoint (if so, either extend said test to cover all points, or skip the next 3 checkboxes)
- [ ] Device-side (C) component developed
- [ ] Bazel build rules developed
- [ ] Host-side component developed
- [ ] HJSON test plan updated with test name (so it shows up in the dashboard)
- [ ] Test added to dvsim nightly regression (and passing at time of checking)
- [ ] For SiVal test cases, test is running relevant FPGA or silicon regression

**Comment by github-gcontributor** [comment]
HI @vogelpi,
Please share your inputs.  Also feel free to add the right POC for the issue.
thanks.

**Comment by vogelpi** [comment]
Hi @github-gcontributor
Sorry for my late response, I was OoO and just returned to work now.

Thanks for reporting the issue. There are different angles to the problem here:
- First of all, what reasonable health test values are is still under discussion (see e.g. #19392) and heavily depends on the technology and the analog noise source.
- On FPGA, we don't  have a real noise source at hand (the AST model implemented on FPGA models the noise source using an LFSR). So we would have to use different health test parameters on FPGA and on the ASIC SiVal target or alternatively disable part of the test on FPGA.
- These tests are probabilistic in nature and besides changing the threshold values, we might also need to tune other settings such as the window size etc.

But let's maybe stick to the threshold values first. IIUC, you're saying that:
- Value 0x1FF never generates an alert
- Value 0x1FA never generates an alert
- Value 0x1F0 never generates an alert
- Value 0x10F never allows to pass the startup health checks

Can you maybe do a sweep for the values between 0x1F0 and 0x10F. I think somewhere in that window there should be a point where we get some failures but where we still pass the startup health checks.

**Comment by github-gcontributor** [comment]
Hi @vogelpi 

Thanks for your suggestions.
I tried sweeping the values and here is my observation.

0x112 | EDN Stuck
0x113 | EDN Stuck
0x114 | EDN Stuck
0x115 | AES, OTBN never generates an alert
0x116 | AES, OTBN never generates an alert
0x117 | AES, OTBN never generates an alert

The recoverable alert does not get generated in these values.

I00080 entropy_src_fips_mode_health_test.c:450] Before Alerts: Test 1: high_watermark=5 low_watermark=0 high_fails=0 low_fails=0 total_fails=0
I00081 entropy_src_fips_mode_health_test.c:453] Alert Fail Counts for Test 1: HighFails=0 LowFails=0
I00082 entropy_src_fips_mode_health_test.c:450] Before Alerts: Test 3: high_watermark=50 low_watermark=0 high_fails=0 low_fails=0 total_fails=0
I00083 entropy_src_fips_mode_health_test.c:453] Alert Fail Counts for Test 3: HighFails=0 LowFails=0
I00084 entropy_src_fips_mode_health_test.c:450] Before Alerts: Test 4: high_watermark=148 low_watermark=107 high_fails=0 low_fails=0 total_fails=0
I00085 entropy_src_fips_mode_health_test.c:453] Alert Fail Counts for Test 4: HighFails=0 LowFails=0
I00086 entropy_src_fips_mode_health_test.c:450] Before Alerts: Test 5: high_watermark=0 low_watermark=65535 high_fails=0 low_fails=0 total_fails=0
I00087 entropy_src_fips_mode_health_test.c:453] Alert Fail Counts for Test 5: HighFails=0 LowFails=0
E00088 entropy_src_fips_mode_health_test.c:520] No recoverable alerts detected when expected

**Comment by vogelpi** [comment]
Thanks for getting back. What platform are you using exactly?

**Comment by github-gcontributor** [comment]
I am using FPGA setup CW340 + Hyperdebug

**Comment by vogelpi** [comment]
Hi @github-gcontributor , I think it makes sense to try this on the FPGA setup but we might need other parameters on the ASIC eventually.

I now had a closer look at the RTL and the documentation. What I wasn't aware of before (but I think it is worth checking) is that the health test threshold registers only accept more stringent value. So say after a reset you first set the high threshold to 0x115 and then try to set it to 0x1F0, then this second write will not go through. You first have to reset the module / the chip. This may have a big impact on how you have to perform the sweep. To be on the safe side, I recommend reading back the health test threshold values after configuration.

One thing that we could try is setting multiple health test thresholds. If I understand correctly, you just set one threshold now. What I would do to start is configuring the one value before the thing doesn't boot anymore (0x115) and then do the same sweeping for another healt test or e.g. the low threshold and see what happens.

**Comment by github-gcontributor** [comment]
Hi @vogelpi,

Hope you are doing good.
After having multiple internal discussions and trying out tests with changes, we have come to certain findings.  
Attaching the the slide deck outlining our investigation into the non-deterministic outcomes of the FIPS Stringent mode test.

As the documentation doesn't provide clear guidance for this scenario, we would greatly appreciate your valuable input and any insights you may have to arrive to a fruitful conclusion.

Thanks.
[OpenTitan - FIPS stringent mode Health Test.pptx](https://github.com/user-attachments/files/18776168/OpenTitan.-.FIPS.stringent.mode.Health.Test.pptx)

**Comment by vogelpi** [comment]
Hi @github-gcontributor ,
thanks for sharing the slides, they are very useful.

Regarding the non-determinism issue: there are two main factors involved here:
1) The noise source: the main impact of this is how long it takes for ENTROPY_SRC to trigger an alert. I would assume this will ultimately vary with temperature. I would recommend to define a window in which a failure is expected. When it exactly happens will always vary a bit with the setup (I would e.g. expect the chip to heat up during a regression run). 

2) The random number consumption by other EDN0 consumers. AES always consumes a fixed number of bits per block (the settings you use make sense by the way), but there are other blocks connected to EDN0 which also use entropy. How much entropy they consume cannot always be controlled by software as there are things that are running automatically in the background like the AST and OTP_CTRL. How much entropy is used by these blocks also depends on the ROM_EXT version that is run as ROM_EXT may enable/lock some countermeasures etc. Also, here one probably has to define an acceptable window and I would expect this window needs to be tuned again when switching to another ROM_EXT. I think you probably want a moderate number of AES iterations to be on the safe side. A reseed interval of 4 sounds reasonable to me.

As for the AES issues: this is unexpected. I've checked your test code, the DIF code and the testutils code involved, this looks all reasonable. Based on the settings you're using and how you program AES, I would only expect to see timeouts when waiting for OUTPUT_VALID. But I wouldn't expect to see timeouts waiting for INPUT_READY or even mismatches between the encrypted and decrypted data. Can you provide some more details please on when exactly you see latter failures?

**Comment by github-gcontributor** [comment]
Hi @vogelpi,

Thank you for reviewing the latest slides and for your follow-up questions. 

As you noted, it does convey to us that the trigger point for ENTROPY_SRC alert vary with temperature and the inherent statistical nature of the noise source. In our runs, when the alert first asserts and the entropy consuming peripheral starves, has some inherent timing difference which would generate atleast a few entorpy source data. Also, other consumers connected to EDN0—such as AST, OTP, etc —will consume random bits automatically in the background. As a result, software cannot fully control how quickly EDN or CSRNG data is generated and depleted, leading to more variation in the number of blocks we can process before we see the starvation.

Regarding, AES Hangs at InputReady or OutputValid, in our logs most timeouts or stalls indeed occur when waiting for OutputValid, as you mentioned. This matches the scenario where no new mask bits are available to complete the encryption/decryption.
However, in some runs with earlier versions of the code (not part of the main code base anymore), we observed timeouts due to a. OutputValid and/or InputReady (did not have sufficient probe then) 
b. Mismatched data (in AES 128-bit mode). I suspect that was due to extreme leftover data starvation: the pipeline did not have enough random bits even to acknowledge the next block’s load. We do not have a consistently reproducible snippet in the current environment, but historically it did appear in some cases.  We would like to know your thoughts on it.

To summarize, based on your guidance and our observations, 
1. we believe that the test’s non-deterministic outcome is expected: it is sensitive to temperature, background consumers, and leftover data in EDN/CSRNG. 
2. The exact iteration at which we starve for entropy is inherently non-deterministic.
3. We see mismatch if no fresh data is available in the pipeline.
We hope this clarifies our findings and if we can close the issue as inherent-non-deterministic test that would not be part of automation (always require to run manually and analyze the results)

Thank you again for your helpful feedback and suggestions.

**Comment by vogelpi** [comment]
Hi @github-gcontributor ,

thanks for the feedback. I am replying inline below.
> 
> To summarize, based on your guidance and our observations,
> 
>     1. we believe that the test’s non-deterministic outcome is expected: it is sensitive to temperature, background consumers, and leftover data in EDN/CSRNG.

Yes, correct.

> 
>     2. The exact iteration at which we starve for entropy is inherently non-deterministic.

Yes, correct.

>     3. We see AES failures due to wait/mismatch if no fresh data is available in the pipeline.> 

Based on your description above, I would assume that the mismatches were due to programming errors. Because the module is extensively verified at the block-level including also the stalling behavior when waiting for entropy. The way the current source code is written (DIF plus test library and your test), the code should only ever timeout when waiting for OUTPUT_VALID. But if the hardware is operated in the most efficient way, software will input multiple blocks shortly after each other (by polling INPUT_READY). If INPUT_READY doesn't go high, this means the current block isn't finishing (due to no entropy) and consequently the module doesn't absorb the current data in the input register which means INPUT_READY remains at 0. If you then still provide new data, you'll get also mismatches. I would assume your previous test implementation triggered some behavior similar like that.

> We hope this clarifies our findings and if we can close the issue as inherent-non-deterministic test that would not be part of automation (always require to run manually and analyze the results)

Your explanations all sound reasonable to me. But instead of running the test manually, I would try to see if there is a parameter set for which the test passes in say 90% or 95% of the cases. Then you can include the test in automated runs. It will fail here and there but over time it should keep passing in most cases. If the regression logs indeed keep showing 100% failures after a few regression runs, on should start an investigation.

**Comment by github-gcontributor** [comment]
Thanks for the valuable feedback.
https://github.com/lowRISC/opentitan/pull/25719 
Kindly review the checked in file to move forward for closure.
